OpenRA.Mods.Cnc/Traits/World/WithResourceAnimation.cs

```diff
@@ -0,0 +1,105 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2021 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common;
+using OpenRA.Mods.Common.Effects;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Cnc.Traits
+{
+	[TraitLocation(SystemActors.World | SystemActors.EditorWorld)]
+	[Desc("Allows to play animations on resources.", "Attach this to the world actor.")]
+	public class WithResourceAnimationInfo : TraitInfo, Requires<IResourceLayerInfo>
+	{
+		[FieldLoader.Require]
+		[Desc("Resource types to animate.")]
+		public readonly HashSet<string> Types = null;
+
+		[Desc("The percentage of resource cells to play the animation on.", "Use two values to randomize between them.")]
+		public readonly int[] Ratio = { 1, 5 };
+
+		[Desc("Tick interval between two animation spawning.", "Use two values to randomize between them.")]
+		public readonly int[] Interval = { 200, 500 };
+
+		[FieldLoader.Require]
+		[Desc("Animation image.")]
+		public readonly string Image = null;
+
+		[SequenceReference(nameof(Image))]
+		[Desc("Randomly select one of these sequences to render.")]
+		public readonly string[] Sequences = new string[] { "idle" };
+
+		[PaletteReference]
+		[Desc("Animation palette.")]
+		public readonly string Palette = null;
+
+		public override object Create(ActorInitializer init) { return new WithResourceAnimation(init.Self, this); }
+	}
+
+	class WithResourceAnimation : IWorldLoaded, ITick
```

Comment:
This probably wants to be `public` as well to match the other classes and ctor visibility?

Reply:
That is usually a good idea for mod code to access it.